### PR TITLE
Support `show`-ing all aspects at once

### PR DESCRIPTION
**Pull request by dominiklohmann**
This makes the aspect in `show <aspect>` optional. When the aspect is not specified, all aspects are used automatically.

image

### CODE REVIEW

File: web/docs/operators/sources/show.md [hunk 1]
```diff
@@ -45,6 +45,8 @@ look at what's going on:
   available from the `/serve` API endpoint.
 - `types`: shows all known types at a remote node.
 
+When no aspect is not specified, all are shown.
```
Comment (jachris) [inline web/docs/operators/sources/show.md:48]:
```suggestion
When no aspect is specified, all are shown.
```

File: web/docs/operators/sources/show.md [hunk 2]
```diff
@@ -72,3 +74,9 @@ Show the version of a remote node:
 ```
 remote show version
 ```
+
+Show all aspects of a remote node:
+
+```
+show
+```
```
Comment (jachris) [inline web/docs/operators/sources/show.md:82]: This nicely illustrates the confusion.

File: libtenzir/builtins/operators/show.cpp [hunk 3]
```diff
@@ -51,7 +62,10 @@ class show_operator final : public crtp_operator<show_operator> {
   }
 
   auto location() const -> operator_location override {
-    return get()->location();
+    if (const auto* plugin = get()) {
+      return plugin->location();
+    }
+    return operator_location::remote;
```
Comment (jachris) [inline libtenzir/builtins/operators/show.cpp:68]: I find this a bit confusing.

**Review by jachris — APPROVED** [review]
+1

File: web/docs/operators/sources/show.md (hunk 2 above)
Comment (dominiklohmann) [reply web/docs/operators/sources/show.md:82]: I think the problem more generally is that an operator should not set its location based on its arguments. Changing this to always be remote would simplify the operator a lot, but it breaks `tenzir 'show version'` without a node—which I think would be fine.

@mavam wdyt?

Comment (mavam) [reply web/docs/operators/sources/show.md:82]: >I think the problem more generally is that an operator should not set its location based on its arguments.

I lack understanding to comment on that, but...

> Changing this to always be remote would simplify the operator a lot, but it breaks tenzir 'show version' without a node—which I think would be fine.

...I find it more helpful to show the version of the `tenzir` binary than the remote node.

In general, for the CLI to be remain flexible, I would argue in the opposite direction: if possible, do *not* require a node execution context.

**Comment by mavam** [comment]
Nice. A great example for the multi-schemaness in our pipelines!

File: web/docs/operators/sources/show.md (hunk 2 above)
Comment (dominiklohmann) [reply web/docs/operators/sources/show.md:82]: The problem is that it's hard to document and to explain the expected behavior to the user. When you use `show`, the version aspect is run remotely. When you use `show version`, it is run locally. If I understood Jannis correctly, then the both of us think that this is a confusing user experience.

Comment (dominiklohmann) [reply web/docs/operators/sources/show.md:82]: Operator locations themselves are already a hard-to-grok topic, but if an operator can dynamically change its location like the `show` operator than it's hard to understand what's going on.

Comment (mavam) [reply web/docs/operators/sources/show.md:82]: I agree that this is confusing, and (now) also that location should be independent of arguments.

I still find it essential to not lose the ability to show the local version. Would anything speak against introducing `--version` to the executable? The expectation is very clear hear: just show the version of the binary.

Comment (jachris) [reply web/docs/operators/sources/show.md:82]: I find it okay-ish that the location depends on the argument. That `show version` and `show` use different locations to retrieve the version is what's confusing to me, and the same goes for every other anywhere/remote aspect. I guess most of the times the result is the same. However, `show operators` for example can now often change with `disable-plugins`.

Comment (dominiklohmann) [reply web/docs/operators/sources/show.md:82]: Are there any aspects other than `version` for which we would need a local alternative?

Comment (jachris) [reply web/docs/operators/sources/show.md:82]: If you use `tenzir` without a node, then you'd probably want `show operators` to show the local operators, for example.

Comment (dominiklohmann) [reply web/docs/operators/sources/show.md:82]: My current intuition is to just merge this as-is. I think as soon as we work on merging results of multiple operators into one (e.g., via nested pipelines) this topic will come up again anyways.

Wdyt, yes/no?

Comment (jachris) [reply web/docs/operators/sources/show.md:82]: Perhaps you just remove the "show the version of a remote node" example and we'll just glance over the fact that this is a bit inconsistent.

Comment (dominiklohmann) [reply web/docs/operators/sources/show.md]: Done and pushed. Resolving this for now.

**Pull request merged**